blog: gpt-oss performance optimization technical deep dive (#159)
vllm-project/vllm-project.github.io: _posts/2026-02-01-gpt-oss-optimizations.md

---
layout: post

> title: "Optimizing across the pareto: a deep-dive on gpt-oss performance optimizations on NVIDIA’s Blackwell GPU."

Reviewer: ```suggestion
title: "GPT-OSS Performance Optimizations on NVIDIA Blackwell: Pushing the Pareto Frontier"
```

author: "The vLLM and NVIDIA team"
image: /assets/figures/blackwell-inferencemax/gpt-oss-120b-8k-1k-nov-jan.png
---

> **TL;DR:** In collaboration with the open-source community, vLLM \+ NVIDIA has achieved significant performance milestones on the `gpt-oss-120b` model running on NVIDIA's Blackwell GPUs. Through deep integration with FlashInfer, novel kernel fusions via `torch.compile`, and various inference runtime features, we have set a new record for the model’s performance Pareto frontier —simultaneously optimizing for maximum throughput (+38%) and best interactivity(+13%). 

Reviewer: ```suggestion
**TL;DR:** In collaboration with the open-source community, vLLM \+ NVIDIA has achieved significant performance milestones on the `gpt-oss-120b` model running on NVIDIA's Blackwell GPUs. Through deep integration with FlashInfer, novel kernel fusions via `torch.compile`, and various inference runtime features, we have set a new record for the model’s performance Pareto frontier —simultaneously optimizing for maximum throughput (+38%) and best interactivity (+13%). 
```

This post details the engineering journey, technical breakthroughs, and instructions to reproduce the results. Continuous benchmarks are also available on **[SemiAnalysis Inference MAX](https://inferencemax.semianalysis.com/) and [vLLM Recipes](https://docs.vllm.ai/projects/recipes/en/latest/OpenAI/GPT-OSS.html)**. 

## Table of Contents

- [Introduction](#introduction)
- [FlashInfer + torch.compile](#fi-tc)
- [Runtime Improvements](#runtime)
- [Deployment Recipes](#recipes)
- [Results](#results)
- [Next Steps](#next-steps)
- [Acknowledgements](#acknowledgements)

---

> ## Introduction:

Reviewer: Remove colon to be consistent
```suggestion
## Introduction
```

> Optimizing for a single metric—like maximum throughput or single-batch latency—is often insufficient for real-world deployments. Different use cases require different latency constraints and request concurrency. As a result, the real challenge lies in optimizing the **Pareto frontier**: the curve that represents the best possible trade-off between **Tokens Per Second (TPS) per GPU** (TCO) and **TPS per User** (interactivity). Pushing this curve upwards and to the right means delivering faster generation for individual users while allowing more users to share the hardware. [SemiAnalysis InferenceMAX](https://inferencemax.semianalysis.com/) has identified this critical need to measure, report and improve performance data for such LLM inference workloads on modern GPUs.

Reviewer 2: ```suggestion
Optimizing for a single metric—like maximum throughput or single-batch latency—is often insufficient for real-world deployments. Different use cases require different latency constraints and request concurrency. As a result, the real challenge lies in optimizing the **Pareto frontier**: the curve that represents the best possible trade-off between **Tokens Per Second (TPS) per GPU** (TCO, total cost of ownership) and **TPS per User** (interactivity). Pushing this curve upwards and to the right means delivering faster generation for individual users while allowing more users to share the hardware. [SemiAnalysis InferenceMAX](https://inferencemax.semianalysis.com/) has identified this critical need to measure, report and improve performance data for such LLM inference workloads on modern GPUs.
```

> One of the key use-cases is serving OpenAI’s `gpt-oss-120b` model, an natively 4-bit quantized (MXFP4) Mixture-of-Experts (MoE) LLM. It has achieved SoTA model accuracy along with strong agentic capabilities. At the recent SemiAnalysis InferenceMAX showcase, vLLM demonstrated its capability to handle this workload efficiently on NVIDIA’s latest Blackwell (B200/GB200) architecture.

Reviewer 2: ```suggestion
One of the key use-cases is serving OpenAI’s `gpt-oss-120b` model, an natively 4-bit quantized (MXFP4) Mixture-of-Experts (MoE) LLM. It has achieved SoTA model accuracy for its size along with strong agentic capabilities. At the recent SemiAnalysis InferenceMAX showcase, vLLM demonstrated its capability to handle this workload efficiently on NVIDIA’s latest Blackwell (B200/GB200) architecture.
```

> One of the key use-cases is serving OpenAI’s `gpt-oss-120b` model, an natively 4-bit quantized (MXFP4) Mixture-of-Experts (MoE) LLM. It has achieved SoTA model accuracy for its size along with strong agentic capabilities. At the recent SemiAnalysis InferenceMAX showcase, vLLM demonstrated its capability to handle this workload efficiently on NVIDIA’s latest Blackwell (B200/GB200) architecture.

Reviewer: ```suggestion
One of the key use-cases is serving OpenAI’s `gpt-oss-120b` model, a natively 4-bit quantized (MXFP4) Mixture-of-Experts (MoE) LLM. It has achieved SoTA model accuracy for its size along with strong agentic capabilities. At the recent SemiAnalysis InferenceMAX showcase, vLLM demonstrated its capability to handle this workload efficiently on NVIDIA’s latest Blackwell (B200/GB200) architecture.
```

> The heart of the optimizations is extreme hardware-software co-design. The NVIDIA B200/GB200 GPUs introduce powerful features like native FP4 TensorCores and 192GB HBM per GPU, which are critical for serving large MoE models like `gpt-oss`. To leverage this hardware fully, vLLM and NVIDIA teams have integrated with **FlashInfer** and adopted a rigorous optimization strategy focusing on kernel fusion, communication overhead reduction, and host-device overlapping.

Reviewer 2: ```suggestion
The heart of the optimizations is hardware-software co-design. The NVIDIA B200/GB200 GPUs introduce powerful features like native FP4 TensorCores and 192GB HBM per GPU, which are critical for serving large MoE models like `gpt-oss`. To leverage this hardware fully, vLLM and NVIDIA teams have integrated with **FlashInfer** and adopted a rigorous optimization strategy focusing on kernel fusion, communication overhead reduction, and host-device overlapping.
```

* **SiluMul  \+ Quantization:** We fused the SiluMul activation function with NVFP4 quantization operations to further reduce latency, details please see [PR23671](https://github.com/vllm-project/vllm/pull/23671).
* **Pad \+ Quant & Finalize \+ Slice:** We are actively rolling out the [fusion passes, PR30647](https://github.com/vllm-project/vllm/pull/30647) for padding/quantization and finalize/slice operations to further streamline the MoE execution path, with an expected 6% performance gain.

As we identify and develop new fused operations, the team will continue to deliver automatic performance gains via this infrastructure.

## Runtime Improvements <a name="runtime"></a>

On next-generation hardware like Blackwell, the GPU is so fast that the CPU (host) often becomes the bottleneck, struggling to dispatch kernels quickly enough to keep the GPU busy. In addition, `prepare\_batch`, request scheduling and sampling logic also require heavy CPU side logic. This "host overhead" manifests as gaps between kernel executions, degrading performance and overall GPU utilization.

> To address this, we implemented both **Async Scheduler** and **Stream Interval** to vLLM that effectively eliminate host-side overhead.

Reviewer 2: ```suggestion
To address this, we implemented both **Async Scheduling** and **Stream Interval** to vLLM that effectively eliminate host-side overhead.
```

> [Async Scheduler](https://github.com/vllm-project/vllm/pull/23569):

Reviewer 2: ```suggestion
[Async Scheduling](https://github.com/vllm-project/vllm/pull/23569):
```

* **Mechanism:** This scheduler decouples the CPU's request scheduling from the GPU's execution. By allowing the CPU to prepare the next batch of requests while the GPU is still processing the current batch, we effectively hide the host overhead.  
* **Impact:** This optimization is crucial for the `gpt-oss` model, particularly in both high-throughput and min-latency scenarios. On more capable GPUs (H200s, B200s, GB200s), you can expect around a 10% performance gain.

> * **Configuration:** Users can enable this feature using the `--async-scheduling` flag, which is now supported in conjunction with many key features such as speculative decoding and structured outputs. 

Reviewer 2: ```suggestion
* **Configuration:** This has been turned on by default in recent vLLM releases. 
```

* **Configuration:** Users can configure this behavior using the `--stream-interval <num_tokens>` argument. The default value is `1` (standard streaming), but increasing this value (e.g., to `10`) is highly effective for reducing host overhead in high-throughput deployments.

## Deployment Recipes <a name="recipes"></a>

To reproduce the optimized performance for `gpt-oss` on Blackwell GPUs (B200/GB200), we recommend the following configurations in your vLLM deployment recipes. They can also be found under [vLLM Recipes page](https://docs.vllm.ai/projects/recipes/en/latest/OpenAI/GPT-OSS.html).

**Recommended Configuration Flags:**

* **Graph Capture:** `FULL_AND_PIECEWISE` mode is recommended to maximize CUDA graph usage for both prefill and decode phases.  
  * `--cudagraph-mode FULL_AND_PIECEWISE # already applied by default since vLLM 0.12.0`  

>   * `--cuda-graph-capture-size 2048`  

Reviewer 2: I'm pretty sure this is also the default.

Reviewer: I think it is just 1024 as the default for gpt-oss

* **Scheduling:**  

>   * `--async-scheduling` (to hide host overhead).  

Reviewer 2: ```suggestion
  * `--async-scheduling` (to hide host overhead). This has already been applied by default as of vllm==0.14.0. 
```

>   * `--api-server-count 20` or `--stream-interval 20`: This helps decouple the HTTP API server overhead from the inference engine, stabilizing performance at high concurrency.  

Reviewer 2: If this is the only non-default flag, we should just mention this one. 
Ideally the least amount of flags = good UX.

* **MoE Backend:**  

>   * Explicitly enable the optimized Cutlass backend for FP8/FP4 MoE to ensure maximum throughput: `VLLM_USE_FlashInfer_MOE_MXFP4_MXFP8=1`.  

Reviewer: ```suggestion
  * Explicitly enable the optimized Cutlass backend for FP8/FP4 MoE to ensure maximum throughput: `VLLM_USE_FLASHINFER_MOE_MXFP4_MXFP8=1`.  
```

## Results

The combined effect of the optimizations has resulted in a significant uptick of performance since the launch of [InferenceMax](https://blog.vllm.ai/2025/10/09/blackwell-inferencemax.html). Notably, a 38% performance increase at max-throughput, and 13% performance increase at min-latency

<p align="center">
<picture>
<img src="/assets/figures/blackwell-inferencemax/gpt-oss-120b-8k-1k-nov-jan.png" width="100%">
</picture><br>
</p>

> Such improvements are not just for a single use-case, but rather across the entire Pareto curve to behefit the vLLM community at large. 

Reviewer: ```suggestion
Such improvements are not just for a single use-case, but rather **across the entire Pareto curve to benefit the vLLM community at large**. 
```

## Next steps

Our work on `gpt-oss` is ongoing. Here is a look at the active engineering tracks to further push the Pareto frontier. The list can also be found in [Issue 30758](https://github.com/vllm-project/vllm/issues/30758).

### Disaggregation

By separating the Prefill stage and the Decode stage on to different GPUs, we can potentially achieve better throughput per GPU. We are currently experimenting with this setup and find the correct configs that achieve better performance.

> ### DEP performance:

Reviewer: ```suggestion
### Data+Expert parallel performance
```

Our projection shows that using DEP2 (Attention DP \+ MoE EP on 2 GPUs) can potentially achieve higher throughput per GPU compared to TP1 and TP2 at the same latency (TPS/user). However, currently the DEP2 performance is worse than TP1/TP2 mainly due to the MoE kernel selection issue. We are actively working on this to resolve it.

> ### Min-latency:

Reviewer: ```suggestion
### Minimum latency performance
```
